### PR TITLE
feature: variable max

### LINKED ISSUE #43: Dynamic max requests
I was looking at the code and it seems that the plugin allows for the specified `max` option to be a function, but when I try to use a function to generate the max value I get an error such as:

> unhandled exception TypeError: The header content contains invalid characters
> at storeHeader (_http_outgoing.js:436:13)
> at ServerResponse._storeHeader (_http_outgoing.js:350:7)
> at ServerResponse.writeHead (_http_server.js:257:8)
> at onSendEnd (/usr/src/app/node_modules/fastify/lib/reply.js:329:7)
> at onSendHook (/usr/src/app/node_modules/fastify/lib/reply.js:282:5)
> at _Reply.Reply.send (/usr/src/app/node_modules/fastify/lib/reply.js:139:3)
> at onIncr (/usr/src/app/node_modules/fastify-rate-limit/index.js:136:12)
> at LocalStore.incr (/usr/src/app/node_modules/fastify-rate-limit/store/LocalStore.js:19:3)
> at Object.preHandler (/usr/src/app/node_modules/fastify-rate-limit/index.js:110:27)
> at hookIterator (/usr/src/app/node_modules/fastify/lib/hooks.js:124:10)

Is there support for a non-static max requests value and I'm missing something or is this not implemented?

In my use case different clients connecting to the API have different max limits and as such I would like to make the max option depend on a function provided by me that would return a value to be checked against.

Thanks for your time!

### PR DESCRIPTION
**Pull request by Eomm**
Close #43 

This lets the user set a max based on:
- timing
- api key
- qos
- all new ideas that could come

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ fastify.register(require('fastify-rate-limit'), {
 ```
 
 - `global` : indicates if the plugin should apply the rate limit setting to all routes within the encapsulation scope
-- `max`: is the maximum number of requests a single client can perform inside a timeWindow.
+- `max`: is the maximum number of requests a single client can perform inside a timeWindow. It can be a sync function with the signature `(req, key) => {}` also, and it **must** return a number
```
Comment (StarpTech) [inline README.md]: I'd also describe whats `key`.

File: index.js [hunk 2]
```diff
@@ -142,10 +143,18 @@ function buildRouteRate (pluginComponent, params, routeOptions) {
         }
 
         res.code(429)
-          .header('X-RateLimit-Limit', params.max)
```
Comment (StarpTech) [inline index.js]: Use the kebab-case notation https://developer.twitter.com/en/docs/basics/rate-limiting it's more common

File: test/global-rate-limit.test.js [hunk 3]
```diff
@@ -347,3 +347,65 @@ test('does not override the preHandler as an array', t => {
...
+    max: (req, key) => { return key === 'pro' ? 3 : 2 },
+    timeWindow: 10000
+  })
+
+  fastify.get('/', (req, res) => { res.send('hello') })
+
+  fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+    t.error(err)
+    t.strictEqual(res.statusCode, 200)
+    fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
```
Comment (StarpTech) [inline test/global-rate-limit.test.js]: That is :boom: 
https://miro.medium.com/max/721/1*Co0gr64Uo5kSg89ukFD2dw.jpeg

File: index.js [hunk 4]
```diff
@@ -142,10 +143,18 @@ function buildRouteRate (pluginComponent, params, routeOptions) {
...
+          .header('X-RateLimit-Limit', maximum)
           .header('X-RateLimit-Remaining', 0)
           .header('Retry-After', params.timeWindow)
```
Comment (StarpTech) [inline index.js]: Use `x-rate-limit-reset` e.g https://developer.twitter.com/en/docs/basics/rate-limiting

File: README.md [hunk 5]
```diff
@@ -85,6 +85,17 @@ fastify.register(require('fastify-rate-limit'), {
 })
 ```
```
Comment (StarpTech) [inline README.md]: I'd also document the headers.

File: test/global-rate-limit.test.js (hunk 3 above)
Comment (Eomm) [reply test/global-rate-limit.test.js]: Yeah 😁 I wrote with async before Travis reminded me that doesn't work on node 6 😅

File: index.js (hunk 4 above)
Comment (Eomm) [reply index.js]: I have another pr coming soon for this header that isn't supported right now

File: test/route-rate-limit.test.js [hunk 6]
```diff
@@ -572,3 +572,61 @@ test('custom error response', t => {
...
+                    t.strictEqual(res.statusCode, 429)
+                  })
+                })
+              })
+            })
+          })
+        })
+      })
+    })
+  })
```
Comment (mcollina) [inline test/route-rate-limit.test.js]: I would refactor this code a bit to avoid so much nesting.

Comment (Eomm) [reply test/route-rate-limit.test.js]: Done

**Pull request merged**